File: backend/algorithms/spaced_repetition.py
# ===== Enhanced SM-2 Spaced Repetition =====
# Based on the SuperMemo SM-2 algorithm with BKT integration
# and exam-date-aware interval compression.

from datetime import datetime, timedelta


def schedule_review(
    current_interval: int,
    current_ease_factor: float,
    new_mastery: float,
    exam_date: str | None = None,
) -> dict:
    """
    Calculate the next review schedule after a practice session.

    Combines SM-2 interval logic with the student's BKT mastery level.
    If an exam date is approaching, intervals are compressed to ensure
    adequate coverage before the exam.

    Args:
        current_interval:    Days until the current review was due (default 1).
        current_ease_factor: SM-2 ease factor (default 2.5).
        new_mastery:         Updated BKT mastery level (0-1) after practice.
        exam_date:           Optional ISO date string for the upcoming exam.

    Returns:
        Dict with keys: interval, easeFactor, nextReviewDate
    """
    interval = current_interval or 1
    ease_factor = current_ease_factor or 2.5

    # ---- Adjust interval and ease factor based on mastery level ----

    if new_mastery >= 0.85:
        # Mastered — extend the review interval
        ease_factor = max(1.3, ease_factor + 0.1)
        if interval == 1:
            interval = 3
        elif interval <= 3:
            interval = 7
        else:
            interval = round(interval * ease_factor)

    elif new_mastery >= 0.6:
        # Proficient — maintain current interval
        interval = max(1, round(interval * 1.0))

    elif new_mastery >= 0.4:
        # Learning — shorten interval slightly
        ease_factor = max(1.3, ease_factor - 0.15)
        interval = max(1, round(interval * 0.7))

    else:
        # Struggling — reset to daily review
        ease_factor = max(1.3, ease_factor - 0.2)
        interval = 1

    # ---- Exam-date compression ----
    # If the exam is sooner than the interval, compress to fit
    if exam_date:
        try:
            exam_dt = datetime.fromisoformat(exam_date.replace("Z", "+00:00"))
            days_to_exam = max(0, (exam_dt - datetime.now(exam_dt.tzinfo)).total_seconds() / 86400)
            if days_to_exam < interval:
                interval = max(1, int(days_to_exam / 3))
        except (ValueError, TypeError):
            pass  # If date parsing fails, skip compression

    # Cap at 30 days
    interval = min(interval, 30)

    next_review_date = (datetime.utcnow() + timedelta(days=interval)).isoformat() + "Z"

    return {
        "interval": interval,
        "easeFactor": ease_factor,
        "nextReviewDate": next_review_date,
    }


def get_due_reviews(mastery_map: dict) -> list[dict]:
    """
    Get concepts that are due for review today.

    A concept is due if:
    - It has a nextReviewDate in the past, OR
    - It has been practiced at least once but was never scheduled.

    Returns:
        List of concept mastery dicts, sorted weakest-first.
    """
    now = datetime.utcnow()
    due = []

    for concept_id, cm in mastery_map.items():
        next_review = cm.get("nextReviewDate")
        if not next_review:
            # Practiced but never scheduled → due
            if cm.get("totalAttempts", 0) > 0:
                due.append(cm)
        else:
            try:
                review_dt = datetime.fromisoformat(next_review.replace("Z", "+00:00"))
                if review_dt <= now.replace(tzinfo=review_dt.tzinfo):
                    due.append(cm)
            except (ValueError, TypeError):
                pass

    # Sort by mastery ascending (weakest concepts first)
    due.sort(key=lambda c: c.get("mastery", 0))
    return due


def get_weakest_concepts(mastery_map: dict, limit: int = 5) -> list[dict]:
    """
    Get the weakest concepts that need priority study.

    Args:
        mastery_map: Dict of concept_id -> mastery data.
        limit:       Maximum number of concepts to return.

    Returns:
        List of up to `limit` concept mastery dicts, sorted weakest-first.
    """
    weak = [
        cm for cm in mastery_map.values()
        if cm.get("mastery", 0) < 0.7
    ]
    weak.sort(key=lambda c: c.get("mastery", 0))
    return weak[:limit]

File: src/lib/mastery/spaced-repetition.ts
// ===== Enhanced SM-2 Spaced Repetition =====
// Based on SuperMemo SM-2 with BKT integration and exam-date awareness

import { ConceptMastery } from "../knowledge-graph/types";

/**
 * Calculate the next review schedule after a practice session.
 * Integrates BKT mastery level with SM-2 interval calculation.
 */
export function scheduleReview(
    conceptMastery: ConceptMastery,
    newMastery: number,
    examDate?: string | null
): Pick<ConceptMastery, "interval" | "easeFactor" | "nextReviewDate"> {
    let interval = conceptMastery.interval || 1;
    let easeFactor = conceptMastery.easeFactor || 2.5;

    if (newMastery >= 0.85) {
        // Mastered — extend interval
        easeFactor = Math.max(1.3, easeFactor + 0.1);
        if (interval === 1) {
            interval = 3;
        } else if (interval <= 3) {
            interval = 7;
        } else {
            interval = Math.round(interval * easeFactor);
        }
    } else if (newMastery >= 0.6) {
        // Proficient — maintain interval
        interval = Math.max(1, Math.round(interval * 1.0));
    } else if (newMastery >= 0.4) {
        // Learning — reduce interval
        easeFactor = Math.max(1.3, easeFactor - 0.15);
        interval = Math.max(1, Math.round(interval * 0.7));
    } else {
        // Struggling — reset to daily
        easeFactor = Math.max(1.3, easeFactor - 0.2);
        interval = 1;
    }

    // Exam-date compression: if exam is close, compress intervals
    if (examDate) {
        const daysToExam = Math.max(
            0,
            (new Date(examDate).getTime() - Date.now()) / (1000 * 60 * 60 * 24)
        );
        if (daysToExam < interval) {
            interval = Math.max(1, Math.floor(daysToExam / 3));
        }
    }

    // Cap interval at 30 days for practical scheduling
    interval = Math.min(interval, 30);

    const nextReviewDate = new Date(
        Date.now() + interval * 24 * 60 * 60 * 1000
    ).toISOString();

    return { interval, easeFactor, nextReviewDate };
}

/**
 * Get concepts due for review today.
 */
export function getDueReviews(
    masteryMap: Record<string, ConceptMastery>
): ConceptMastery[] {
    const now = new Date();
    return Object.values(masteryMap)
        .filter((cm) => {
            if (!cm.nextReviewDate) return cm.totalAttempts > 0; // practiced but never scheduled
            return new Date(cm.nextReviewDate) <= now;
        })
        .sort((a, b) => a.mastery - b.mastery); // weakest first
}

/**
 * Get concepts that are weakest (for priority study).
 */
export function getWeakestConcepts(
    masteryMap: Record<string, ConceptMastery>,
    limit: number = 5
): ConceptMastery[] {
    return Object.values(masteryMap)
        .filter((cm) => cm.mastery < 0.7)
        .sort((a, b) => a.mastery - b.mastery)
        .slice(0, limit);
}
